torch/csrc/autograd/functions/convolution.cpp

```diff
@@ -120,204 +143,432 @@ static auto view3d(const Tensor& tensor) -> std::unique_ptr<Tensor> {
+        } else {
+          tensor_list outputs(groups);
+          for (int g = 0; g < groups; ++g) {
+            auto input_g = subtensor(input.get(), 1, groups, g);
+            auto weight_g = subtensor(weight.get(), 0, groups, g);
+            auto bias_g = subtensor(bias.get(), 0, groups, g);
+            outputs[g] = compute_output(
+                input_g.get(), weight_g.get(), bias_g.get(),
+                columns[g].get(), ones[g].get(), kernel_size, *this, false);
+          }
+          output = cat(outputs, 1);
+        }
+      }
 
-  std::unique_ptr<Tensor> output;
-  tensor_list columns(groups);
-  tensor_list ones(groups);
-  std::unique_ptr<Convolution> convolution;
+      if (k == 3) {
+        output = view3d(*output);
+      }
 
-  if (use_cudnn(*input)) {
-#ifdef WITH_CUDNN
-    output = input->newTensor();
-    output->resize(output_size(*input, *weight));
-    if (transposed) {
-      convolution.reset(cudnn_convolution_transpose_full_forward(
-          state, torch::cudnn::getCudnnHandle(), torch::cudnn::getCudnnDataType(*input),
-          (THVoidTensor*)input->cdata(), (THVoidTensor*)weight->cdata(),
-          bias ? (THVoidTensor*)bias->cdata() : nullptr, (THVoidTensor*)output->cdata(),
-          padding, stride, dilation, groups, benchmark));
-    } else {
-      convolution.reset(cudnn_convolution_full_forward(
-          state, torch::cudnn::getCudnnHandle(), torch::cudnn::getCudnnDataType(*input),
-          (THVoidTensor*)input->cdata(), (THVoidTensor*)weight->cdata(),
-          bias ? (THVoidTensor*)bias->cdata() : nullptr, (THVoidTensor*)output->cdata(),
-          padding, stride, dilation, groups, benchmark));
-    }
-#endif
+      auto outputs = as_tensor_list(std::move(output));
+      std::cout << "Created Backward convolution" << convolution.get() << std::endl;
+      return wrap_outputs(inputs, std::move(outputs), [&](FunctionFlags f) {
+        return std::make_shared<ConvBackward>(
+            f, *this,
+            inputs[0]->save(this), inputs[1]->save(this), Variable::save_opt(inputs[2].get(), this),
+            std::move(columns), std::move(ones), convolution);
+      });
   } else {
-    for (int g = 0; g < groups; ++g) {
-      columns[g] = input->newTensor();
-      ones[g] = input->newTensor();
-    }
-    if (groups == 1) {
-      output = compute_output(
-          input.get(), weight.get(), bias.get(),
-          columns[0].get(), ones[0].get(), kernel_size, *this);
-    } else {
-      tensor_list outputs(groups);
-      for (int g = 0; g < groups; ++g) {
-        auto input_g = subtensor(input.get(), 1, groups, g);
-        auto weight_g = subtensor(weight.get(), 0, groups, g);
-        auto bias_g = subtensor(bias.get(), 0, groups, g);
-        outputs[g] = compute_output(
-            input_g.get(), weight_g.get(), bias_g.get(),
-            columns[g].get(), ones[g].get(), kernel_size, *this);
+      std::cout << "ConvForward in grad mode" << std::endl;
```

Comment:
Because the convForward in forward mode which means created by the python module and convForward in grad mode which means created by ConvBackward will perform different actions